[충남대 FE 김은수] 2단계 주문하기 페이지 구현하기

shushuburger opened the pull request:
안녕하세요! 충남대 FE 김은수입니다!!

### 요구사항

- [x]  본인만의 기준으로 일관된 코드를 작성해주세요.
- [x]  기능 단위로 나누어 커밋을 해주세요.
- [x]  재사용 가능성이 높은 컴포넌트, Hooks가 보이면 재사용성을 고려하여 만들어요.
- [x]  별도의 From, Validation 라이브러리를 사용하지 않고 useState로만 구현해요.
- [x]  로그인 정보
    - [x]  로그인 정보를 어디서든 사용할 수 있게 해요.
    - [x]  새로고침을 하더라도 로그인 정보가 유지되도록 해요.
    - [x]  로그인 한 유저는 로그인 버튼 클릭 시 마이페이지(**`/my`**)로 연결돼요.
    - [x]  마이페이지는 로그인 한 사람만 접근 가능해요. 로그인 하지 않은 사람이 접속하면 로그인 페이지로 연결해요.
    - [x]  마이페이지에서 로그아웃을 할 수 있어요. 로그아웃 시 로그인(**`/login`**) 페이지로 연결돼요.
- [x]  주문하기
    - [x]  상품 아이템을 클릭하면 주문하기 페이지로 연결돼요.
    - [x]  만약 로그인이 되어있지 않다면 로그인을 유도하고, 로그인 성공 시 주문하기 페이지로 접근 가능하게 해요.
    - [x]  [[참고 URL](https://kakaotech-mission2-order-step2.pages.dev/)](https://kakaotech-mission2-order-step2.pages.dev/)을 살펴보고 주문하기에 필요한 기능을 구현해요.
    - [x]  버튼 클릭 시 아래 유효성 검사 조건이 충족되지 않으면 각 폼 필드에 안내 메시지를 보여줘요.
        - [x]  메시지는 반드시 입력 되어야 해요.
        - [x]  보내는 사람 이름이 반드시 입력 되어야 해요.
        - [x]  받는 사람 이름이 반드시 입력 되어야 해요.
        - [x]  받는사람 전화번호가 반드시 입력되고 전화번호 규칙에 맞아야 해요. (01012341234)
        - [x]  수량은 1개 이상이어야 해요.

shushuburger committed:
docs(readme): 2단계 과제 요구사항 기록

shushuburger committed:
feat(auth): 로그인 시 sessionStorage에 사용자 정보 저장 및 리디렉션 처리

- AuthContext 생성하여 login, logout 함수 및 user 상태를 관리하도록 했습니다.
- login 시 'kakaotech/userInfo' 키로 sessionStorage에 이메일을 저장하도록 했습니다.
- 새로고침 시 sessionStorage에 저장되어 있는 내용이 유지되는 것을 확인했습니다.
- useAuth 훅으로 전역에서 정보에 쉽게 접근할 수 있도록 발판을 마련해두었습니다.
- LoginFormSection에서 login() 호출 시 사용자 정보 저장 후 이동하도록 했습니다.
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
feat(page): 로그인한 사용자 전용 MyPage 컴포넌트 생성

- 로그인한 사용자의 email을 표시하는 부분을 만들었습니다.
- 로그아웃 버튼 클릭 시 sessionStorage의 내용을 제거 및 /login으로 이동하도록 했습니다.

shushuburger committed:
feat(navigation): 로그인 여부에 따라 상단 아이콘 버튼 동작 구분

- 로그인 상태이면 아이콘 클릭 시에 /my로 이동하도록 했습니다.
- 로그인 상태가 아니면 로그인 페이지(/login)로 이동하도록 했습니다.
- useAuth 훅으로 로그인 상태 확인하여 동작을 구분하였습니다.
- routes.ts에 MY라는 이름으로 라우팅을 등록해주었습니다.
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
feat(friend-select): 로그인된 사용자의 이메일 아이디를 배너 버튼에 동적 표시

- sessionStorage에 저장된 사용자 이메일에서 아이디를 추출했습니다.
- "{아이디}님! 선물할 친구를 선택해 주세요." 형식으로 버튼 텍스트를 표시하도록 했습니다.
- 로그인하지 않은 경우엔 기존 문구를 그대론 출력하도록 연산자를 사용했습니다.
- README에 로그인 부분 구현 완료를 표시했습니다.

shushuburger committed:
feat(data): 주문하기 메시지 카드용 mock data 추가

- messageCards.ts에 메시지 카드 썸네일/이미지/mock 메시지 배열을 정의했습니다.

shushuburger committed:
feat(route): 주문하기 페이지를 위한 /order/:id 경로 및 OrderPage 컴포넌트 등록

- constants/routes.ts에 ORDER 경로를 추가했습니다.
- 기본적인 요소들로 OrderPage.tsx를 구현해놓았습니다. 수정 예정입니다.

shushuburger committed:
feat(product-card): 상품 카드 클릭 시 해당 상품의 주문 페이지로 이동 처리

- navigate(`/order/{id}`)를 통해 주문 페이지에 진입하도록 했습니다.
- Card 컴포넌트 전체에 클릭 이벤트를 다룰 수 있도록 onClick을 사용해주었습니다.
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
fix(login): 로그인 후 원래 위치로 올바르게 이동하도록 수정

- 로그인 전에 접근했던 경로(`/order/:id`)를 login 페이지의 location.state로 전달하였습니다.
- login 성공 시 해당 경로로 navigate하도록 작성해주었습니다.
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
feat(order): 주문 페이지의 기본 UI 레이아웃 및 스타일 구현

- 메시지 카드 선택 영역, 프리뷰 카드, 텍스트 입력창 UI를 구현했습니다.
- 보내는 사람, 받는 사람 입력 영역에 스타일을 적용해주었습니다.
- refactor 과정을 통해 컴포넌트 분리를 진행할 예정입니다.

shushuburger committed:
refactor(order): CardSelector 컴포넌트 분리 및 적용

- 메시지 카드 선택 영역(CardSlider + PreviewCard)을 CardSelector 컴포넌트로 분리했습니다.
- OrderPage에서 CardSelector 적용 및 기존 코드 제거하였습니다.

shushuburger committed:
refactor(order): MessageInput 컴포넌트 분리 및 적용

- 메시지 카드에 따른 기본 메시지를 보여주는 입력 영역을 MessageInput으로 분리했습니다.
- 스타일 및 텍스트 표현을 기존과 동일하게 유지했습니다.
- OrderPage에서 컴포넌트를 알맞게 적용하였습니다.

shushuburger committed:
refactor(order): OrderPage 주요 UI 컴포넌트 분리

- 메시지 카드 선택 영역을 CardSelector로 분리했습니다.
- 카드 미리보기를 CardPreview로 분리했습니다.
- 메시지 입력창을 MessageInput으로 분리했습니다.
- 보내는 사람 입력폼을 SenderForm으로 분리했습니다.
- 받는 사람 입력폼을 ReceiverForm으로 분리했습니다.
- 상품 정보 표시를 ProductInfo로 분리했습니다.
- 주문 버튼을 OrderSubmitButton으로 분리했습니다.

shushuburger committed:
feat(order): OrderPage에 컴포넌트 및 common 적용 + 유효성 검사 로직 구현

- Button, InputField 등 common 컴포넌트 활용하였습니다.
- 보내는 사람 / 받는 사람 / 전화번호 / 수량 유효성 검사 로직을 구현하였습니다.
- 버튼 클릭 시 모든 필드에 대한 오류 메시지를 표시하도록 했습니다.
- 수량 변경 시 총 주문 금액 실시간으로 반영하도록 했습니다.
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
feat(order): 주문 완료 알림 및 메인 페이지 이동 로직 구현

- 주문 정보가 유효한 경우 alert로 주문 내역(상품명, 수량, 발신자, 메시지) 표시하도록 했습니다.
- 확인 클릭 시 navigate('/')를 통해 홈으로 이동하도록 해주었습니다.
- README에 구현 완료한 내용들을 표시했습니다.

shushuburger committed:
refactor(order): 주문 폼 에러 메시지 및 유효성 검사용 상수 분리

- PHONE_REGEX, MIN_QUANTITY, ]ERROR_MESSAGES를 constants/orderForm.ts로 분리했습니다.
- OrderPage의 유효성 검사 로직에서 문자열 대신 상수 참조하도록 리팩토링해주었습니다.

shushuburger committed:
refactor(order): OrderPage 상태 및 로직 useOrderForm 훅으로 분리

- 메시지, 카드 선택, 제출 처리 등 전체 상태 로직을 hook으로 분리하여 주었습니다.
- 발신자/수신자 정보, 수량, 유효성 검사는 useOrderForm 훅에서 책임지도록 분리했습니다.
- OrderPage.tsx는 렌더링 전용 컴포넌트로 단순화해주었습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
refactor(hook): 1단계 때 남긴 피드백 반영

- email, password를 userInfo로 묶어서 처리했습니다.
- useOrderForm.ts에도 공식 문서를 참고하여 form에 들어가는 value와 error들을 묶어서 관리하게 했습니다.

shushuburger commented:
1단계에서 남아있던 피드백인 "userInfo로 통합해보기"를 수행했습니다.
이번에 주문하기 화면을 만들면서 정말 많은 요소들을 다루게 되어서 "묶어서 관리해야 하는 것"의 필요성을 느꼈던 것 같습니다.

```
  const [person, setPerson] = useState({
    firstName: 'Barbara',
    lastName: 'Hepworth',
    email: 'bhepworth@sculpture.com'
  });

  function handleFirstNameChange(e) {
    setPerson({
      ...person,
      firstName: e.target.value
    });
  }
```

해당 형태를 바탕으로 gpt의 도움을 받아 값들을 묶어서 관리하고 handle 함수에서도 이전 값을 사용할 수 있도록 작업했습니다.

https://react.dev/learn/updating-objects-in-state

shushuburger committed:
refactor(hook): useOrderForm에 message 요소 추가

- 카드에 포함된 메시지도 Form을 제출할 때 검사 대상에 속하기에 hook에 포함시켰습니다.

shushuburger commented on src/components/common/Button.tsx:
```diff
@@ -7,6 +7,7 @@ interface ButtonProps {
   onClick?: () => void;
   backgroundColor?: string;
   height?: string;
+  borderRadius?: string;
```
이번에 로그인 화면에서 사용된 "로그인 버튼"에 이용한 Button.tsx라는 common component를 주문하기 화면에서 "주문하기" 버튼에 이용해보았습니다!!

이때 로그인 화면은 "버튼이 둥글다"라는 특징, 주문하기 화면은 "버튼이 각졌다"라는 특징이 있어 저번 피드백에서 알려주셨던 InputHTMLAttributes를 사용해보려고 했습니다!
그렇지만 InputHTMLAttributes는 html 속성을 props로 제공하고 있어 border-radius 같은 css 속성은 추가적으로 지금 방식처럼 작업을 해줘야 한다고 했습니다.

제가 이해한 내용에 잘못된 점이 없는지 궁금합니다!

Leejin-Yang replied:
기본 버튼의 속성을 가져오고 커스텀 prop을 추가하기 위해서 extends를 사용할 수 있어 보여요.

```
interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  borderRadius?: string;
  ...
}
```

이번 기회에 상속에 대해서도 학습해보시는걸 추천드립니다 😊

shushuburger committed:
refactor(login): 불필요한 코드 제거

Leejin-Yang commented on src/components/LoginFormSection.tsx:
```diff
@@ -5,40 +5,42 @@ import { LOCATION_STATE_KEYS } from '@/constants/navigationState';
   const location = useLocation();
-  const from = location.state?.[LOCATION_STATE_KEYS.FROM] ?? ROUTES.HOME;
+
+  type FromState = {
+    pathname: string;
+    search?: string;
+  };
```
타입은 컴포넌트 외부에서 작성해주세요 😊

shushuburger replied:
이동시켰습니다! 감사합니다

Leejin-Yang commented on src/components/Navigation.tsx:
```diff
@@ -41,11 +43,11 @@ const Navigation = ({ showLoginButton = true }: NavigationProps) => {
-              state: { from: location },
-            })
+            user
+              ? navigate('/my')
+              : navigate('/login', { state: { from: location } })
```
상수로 작성한 경로를 사용할 수 있어보여요 😊

shushuburger replied:
상수를 이용해주었습니다!

Leejin-Yang commented on src/components/OrderSection/ProductInfo.tsx:
```diff
@@ -0,0 +1,68 @@
+
+export default ProductInfo;
+
+const Wrapper = styled.div``;
```
빈 스타일드 컴포넌트는 삭제해도 좋아보여요 😊

shushuburger replied:
차후 너비 조절 과정에서 margin, padding을 조절할 일이 생겨 style을 작성해주었습니다.

```
const Wrapper = styled.div`
  margin: ${({ theme }) => theme.spacing[2]} ${({ theme }) => theme.spacing[0]};
  padding: ${({ theme }) => theme.spacing[4]};
`;
```

Leejin-Yang commented on src/components/OrderSection/SenderForm.tsx:
```diff
@@ -0,0 +1,44 @@
+const SenderForm = ({ value, onChange, error }: SenderFormProps) => {
+  return (
+    <Wrapper>
+      <Label>보내는 사람</Label>
+      <InputField
```
작은 팁을 드리자면, `label` 태그를 활용해 볼 수 있어보여요. 사용하면 클릭 시 연결되어 있는 input에 포커스를 줄 수 있는 기능이 있어요 😊

https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/label

shushuburger replied:
label에 대해 더 학습해보겠습니다!! 알려주셔서 감사합니다!

Leejin-Yang commented on src/components/RankingSection/ProductCard.tsx:
```diff
@@ -1,14 +1,34 @@
+        state: {
+          from: {
+            pathname: `/order/${rank}`,
+            search: '',
```
여기서도 선언한 상수를 사용할 수 있어보이네요

shushuburger replied:
상수 사용해주었습니다!

Leejin-Yang commented on src/hooks/useOrderForm.ts:
```diff
@@ -0,0 +1,144 @@
+    }
+    setFormErrors(prev => ({ ...prev, textMessage: '' }));
+    return true;
+  };
```
validator 함수도 위의 handleChange처럼 필드명을 인자로 받아 처리할 수 있도록 일반화해보는 건 어떨까요?

이렇게 하면 각 필드마다 별도의 validator 함수를 만들지 않아도 되고, 하나의 공통 로직으로 재사용할 수 있어서 더 깔끔하게 관리할 수 있어보여요 😊

shushuburger replied:
말씀해주신 방향으로 수정해보았습니다!!
완전히 모든 로직을 통일하기엔 비교해야 되는 조건들이 다른 것들이 있어 field에 따라서 각각에 맞는 로직을 찾아갈 수 있도록 case문을 사용해 보았습니다.

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,121 @@
+    validateQuantity,
+    validateForm,
+    totalPrice,
+  } = useOrderForm(product.price.sellingPrice);
```
[중요한 개념] 리액트에서 hook은 반드시 컴포넌트 최상단에서 호출되어야 합니다! return문 뒤에 훅을 사용하면 호출 순서가 달라질 수 있어 이슈가 발생핼 수 있어요. 아래 문서 참고하면서 훅은 어떻게 사용해야 할지 학습해보는걸 추천드려요 😊

https://react.dev/warnings/invalid-hook-call-warning

shushuburger replied:
React는 **함수 컴포넌트를 호출**하면서 **Hook을 “순서대로” 처리**하는 특징을 가졌기에 컴포넌트 내부 코드를 **위에서 아래로** 순차적으로 읽고 **Hook 호출 순서에 따라 내부 상태를 관리**한다는 것을 상기했습니다!

return 밑에 hook이 나와버리면 렌더링 시 호출되지 않을 수 있다는 문제가 있습니다.
이렇게 되면 hook이 꼬이면서 보내주신 링크에 있는 Invalid hook call 오류가 발생할 수 있다고 이해했습니다.

그에 useState, useOrderForm과 같이 hook 과 관련된 코드를 상단으로 옮겼습니다.

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,121 @@
+          <SenderForm
+            value={formValues.senderName}
+            onChange={e => {
+              handleChange('senderName', e.target.value);
```
각 필드에 onChange 핸들러를 인라인으로 작성하고 있는데, 하나의 공통 핸들러 함수로 추상화해보는 것도 좋아보여요! 팁을 드리자면, name이나 dataset 속성을 활용할 수 있겠네요 😊

shushuburger replied:
name 속성을 활용해서 공통으로 처리할 수 있도록 해주었습니다!
그 과정에서 ReceiverForm.tsx에 존재하던 기존의 "받는 사람 이름"을 나타내던 name을 조금 더 적절하게 receiverName으로 변수명 변경해주엇습니다.

```
          <SenderForm
            name="senderName"
            value={String(formValues.senderName)}
            onChange={handleInputChange}
            error={formErrors.senderName}
          />
```

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,121 @@
+            quantityError={formErrors.quantity}
+          />
+          <ProductInfo product={product} />
+          <OrderSubmitButton amount={totalPrice} onClick={handleSubmit} />
```
주문 폼도 form 태그를 활용할 수 있어보이네요 😊

shushuburger replied:
form 태그로 감싸주었습니다!!

```
          <OrderSubmitButton amount={totalPrice} />
        </Form>

const Form = styled.form`
  display: flex;
  flex-direction: column;
  gap: ${({ theme }) => theme.spacing[6]};
  margin-top: ${({ theme }) => theme.spacing[5]};
`;
```

Leejin-Yang commented on src/components/OrderSection/OrderSubmitButton.tsx:
```diff
@@ -0,0 +1,36 @@
+  onClick?: () => void;
+}
+
+const OrderSubmitButton = ({ amount, onClick }: OrderSubmitButtonProps) => {
```
버튼 너비가 살짝 안 맞아보이는데, 확인 부탁드려요 😊

<img width="391" alt="Screenshot 2025-07-07 at 18 16 12" src="https://github.com/user-attachments/assets/4c3f6a73-d5e2-4f4e-a8cf-aa061db5edce" />

shushuburger replied:
너비 수정 완료했습니다!

Leejin-Yang commented on src/components/OrderSection/OrderSubmitButton.tsx:
```diff
@@ -0,0 +1,36 @@
+  return (
+    <OrderSubmitButtonWrapper>
+      <Button
+        backgroundColor="#FFE812"
```
작성한 색상 토큰을 사용할 수 있어보여요

shushuburger replied:
토큰 사용 방식으로 변경해주었습니다!

Leejin-Yang commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,47 @@
+  logout: () => void;
+};
+
+const SESSION_KEY = 'kakaotech/userInfo';
```
👍

Leejin-Yang commented on src/hooks/useOrderForm.ts:
```diff
@@ -0,0 +1,144 @@
+  ERROR_MESSAGES,
+} from '@/constants/validation';
+
+type FormFields =
+  | 'senderName'
+  | 'receiverName'
+  | 'receiverPhone'
+  | 'quantity'
+  | 'textMessage';
```
이 필드들은 아래 FormValues의 key 같은데 keyof 키워드를 활용할 수 있어보여요

shushuburger replied:
```
type FormValues = {
  senderName: string;
  receiverName: string;
  receiverPhone: string;
  quantity: number;
  textMessage: string;
};

type FormErrors = Record<keyof FormValues, string>;
type FormField = keyof FormValues;
```

다음과 같이 keyof를 활용해서 조금더 간결한 코드를 작성해보았습니다!

Leejin-Yang commented on src/hooks/useOrderForm.ts:
```diff
@@ -0,0 +1,144 @@
+  textMessage: string;
+}
+
+interface FormErrors {
+  senderName: string;
+  receiverName: string;
+  receiverPhone: string;
+  quantity: string;
+  textMessage: string;
+}
```
모두 스트링이면 [Record](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type)를 활용할 수 있어보이네요!
시간 나실 때 유틸리티 타입도 쓱 보시는걸 추천드립니다 😊

shushuburger replied:
Record에 대해 알아보았습니다!
- 객체의 key와 value 타입을 지정하는 타입
- 동일한 구조를 반복적으로 만들 때 유용

대표적으로 이 2가지 특징이 있다는 것을 알았고 Record를 적용하여서 타입의 안정성을 높였습니다!

```
type FormErrors = Record<keyof FormValues, string>;
type FormField = keyof FormValues;
```

Leejin-Yang commented on src/hooks/useOrderForm.ts:
```diff
@@ -0,0 +1,144 @@
+}
+
+export const useOrderForm = (unitPrice: number) => {
+  const [formValues, setFormValues] = useState<FormValues>({
+    senderName: '',
+    receiverName: '',
+    receiverPhone: '',
+    quantity: 1,
+    textMessage: '',
+  });
```
상태 초기값들도 상수로 관리하는 것도 추천드립니다 😊

shushuburger replied:
```
const DEFAULT_FORM_VALUES: FormValues = {
  senderName: '',
  receiverName: '',
  receiverPhone: '',
  quantity: 1,
  textMessage: '',
};

const DEFAULT_FORM_ERRORS: FormErrors = {
  senderName: '',
  receiverName: '',
  receiverPhone: '',
  quantity: '',
  textMessage: '',
};
```

다음과 같이 상수에 저장을 해주었습니다.

Leejin-Yang reviewed (changes requested):
은수님 안녕하세요 👋
2단계는 요구사항이 많아서 힘드셨을 것 같은데, 잘 진행해주셨네요!
구조적으로 일관되게 잘 진행하신걸로 보여요 👍

몇가지 코멘트 남겼는데, 리팩터링 개념과 하나의 의견으로 봐주시고 반영할 부분 있다면 반영해주세요
그리고 궁금한 점은 언제든지 코멘트와 DM 남겨주시고, 다시 요청 부탁드립니다 😊

---

이건 이번 단계 외적인 부분인데, 홈 화면에 상품 그리드가 깨져 있어서 이 부분 확인 부탁드려요

<img width="320" alt="Screenshot 2025-07-07 at 18 16 35" src="https://github.com/user-attachments/assets/830c864f-4b1b-4529-82e4-8c9e88098ad9" />

shushuburger committed:
fix(layout): 모바일에서 상품 그리드 레이아웃 깨짐 문제 수정

- CardGrid의 grid-template-columns를 minmax(0, 1fr)로 변경하여 콘텐츠 넘침을 방지해주었습니다.
- 모바일에서도 3열 배치가 깨지지 않는 것을 확인했습니다.

shushuburger committed:
fix(order): React Hook 호출 위치 오류 수정

- useOrderForm, useState, useEffect 등 모든 Hook을 OrderPage 컴포넌트 최상단으로 이동시켰습니다.
- return문 아래에 Hook이 존재해 React 렌더링 순서가 깨질 가능성을 제거해주었습니다.

shushuburger committed:
refactor(order): useOrderForm 커스텀 훅 구조 개선 및 의존 코드 정리

- formValues와 formErrors를 Record<FormField, string | number> 형태로 통일해주었습니다.
- 중복된 유효성 검사 함수들을 validateField(field) 하나로 통합하여 조금 더 쉽게 관리하도록 했습니다.
- 초기값을 DEFAULT_FORM_VALUES, DEFAULT_FORM_ERRORS로 상수로 분리하였습니다.

shushuburger committed:
refactor(order): FormField 타입을 keyof FormValues로 변경

- FormValues의 key를 기반으로 FormField 및 FormErrors 을 구현했습니다.

shushuburger committed:
feat(form): OrderPage에 form 태그 및 submit 처리 적용

- 입력 영역을 form 태그로 감싸고 handleSubmit에서 기본 제출 방지 처리를 해주었습니다.

shushuburger committed:
feat(form): name 속성을 활용한 공통 인풋 핸들러 적용

- OrderPage 및 LoginFormSection에서 개별 필드 onChange 핸들러를 하나의 공통 핸들러로 통합했습니다.
- input 및 textarea 요소에 name 속성을 추가하여 handleInputChange에서 식별하도록 했습니다.
- ReceiverForm, SenderForm, MessageInput 등 관련 컴포넌트에 name props 추가해주었습니다.
- useOrderForm의 handleChange, validateField 호출 시 name 기반으로 처리하도록 했습니다.
- LoginFormSection도 이메일/비밀번호 인풋을 name 기반으로 리팩토링해주었습니다.
- 기존에 겹치던 변수명을 변경해주었습니다.

shushuburger committed:
refactor(product): 상수 적용 및 코드 정리

- 라우팅 경로의 경우 상수 처리해주었습니다.
- 불필요한 Wrapper 코드를 삭제해주었습니다.
- 타입 선언 코드의 위치를 변경해주었습니다.

shushuburger committed:
style(button): 색상 토큰 적용 및 버튼 너비/모서리 스타일 조정

- backgroundColor를 theme.color.semantic.kakaoYellow 토큰으로 대체해주었습니다.
- 요소들의 margin, padding을 조절하여 너비를 맞추었습니다.

shushuburger committed:
style(line): 코드 사이 공백 추가

shushuburger commented:
label과 상속에 대해서는 조금 더 학습해보도록 하겠습니다!
f12를 통해 "모바일 뷰"를 볼 수 있다는 사실을 멘토님 덕분에 알게 되었습니다! 앞으로는 작업할 때 해당 상태도 고려하며 작업하겠습니다.

Leejin-Yang commented on src/components/LoginFormSection.tsx:
```diff
@@ -5,40 +5,47 @@ import { LOCATION_STATE_KEYS } from '@/constants/navigationState';
-    validatePassword(value);
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    handleChange(name as 'email' | 'password', value);
```
타입 단언을 사용하지 않고 타입 좁혀보는 걸 추천드려요 😊
어떤 방법이 있을까요?

shushuburger replied:
타입 가드 함수를 통해 조금 더 안전한 방식으로 '타입 좁히기'를 시도했습니다!

```
const isLoginField = (name: string): name is 'email' | 'password' => {
  return name === 'email' || name === 'password';
};

if (isLoginField(name)) {
  handleChange(name, value);
}
```

다음과 같이 작성해보았습니다. ===으로 비교하는 것도 연습했습니다

Leejin-Yang commented on src/hooks/useLoginForm.ts:
```diff
@@ -2,31 +2,44 @@ import { useState } from 'react';
+
+  const handleChange = (field: 'email' | 'password', value: string) => {
+    setUserInfo(prev => ({ ...prev, [field]: value }));
+    if (field == 'email') {
```
안전한 비교를 위해 `===` (strict equal)을 사용해주세요 😊 습관을 들이면 좋아보입니다

shushuburger replied:
변경했습니다!! 앞으로 습관을 들이도록 하겠습니다

Leejin-Yang commented on src/hooks/useOrderForm.ts:
```diff
@@ -0,0 +1,132 @@
+  textMessage: string;
+};
+
+type FormErrors = Record<keyof FormValues, string>;
+type FormField = keyof FormValues;
```
여기서도 작성한 FormField 타입을 활용할 수 있겠네요 😊

```suggestion
type FormErrors = Record<FormField, string>;
type FormField = keyof FormValues;
```

shushuburger replied:
제안해주신 방식을 반영해보았습니다! 감사합니다.

Leejin-Yang commented on src/pages/Mypage.tsx:
```diff
@@ -0,0 +1,58 @@
+import styled from '@emotion/styled';
+import Navigation from '@/components/Navigation';
+
+const MyPage = () => {
```
마이페이지 로그인 하지 않아도 접근이 가능하네요!
이 부분 한번 확인해주세요

인증된 사용자만 들어올 수 있는 경로들은 PrivateRoute로 감싸는 방법도 찾아보면 좋아보입니다 😊

shushuburger replied:
src/routes에 PrivateRoute.tsx를 만들어주어 주소창에 바로 /my, /order/id를 입력하더라도 로그인 여부에 따라 접속 가능 여부가 달라지도록 해주었습니다!

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,119 @@
+  const { id } = useParams();
+  const navigate = useNavigate();
+  const product = mockProducts[Number(id) - 1];
+  if (!product) return <div>잘못된 접근입니다.</div>;
```
아래 커밋에서 수정한 부분이 다시 돌아온걸로 보여요.
확인 부탁드려요 😊

[c71f37d](https://github.com/next-step/react-gift-order/pull/149/commits/c71f37d534a0adb9c2722c4fec6d1e6e4aa78923)

shushuburger replied:
헉...! 이런 실수를 하다니... 수정했습니다!!

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,119 @@
+
+    if (formErrors[name as keyof typeof formErrors]) {
+      validateField(name as any);
+    }
```
이 부분 타입 개선에 도전해보면 좋아보여요!
FormField 타입을 만들어두었으니 활용해보는건 어떨까요?

shushuburger replied:
타입 단언을 방지하기 위해 FormField 타입인지 확인하는 함수를 통하도록 로직을 수행해주었습니다.

Leejin-Yang commented on src/hooks/useLoginForm.ts:
```diff
@@ -2,31 +2,44 @@ import { useState } from 'react';
+    const error = getPasswordError(userInfo.password);
+    setErrors(prev => ({ ...prev, password: error }));
+    return !error;
   };
```
로그인 validator도 다른 작성하신 함수들처럼 하나의 함수로 만들어볼 수 있겠네요 😊

shushuburger replied:
useOrderForm.tsx에서 했던 방식대로 수정해주었습니다!

Leejin-Yang commented on src/components/OrderSection/ReceiverForm.tsx:
```diff
@@ -0,0 +1,66 @@
+import InputField from '@/components/common/InputField';
+
+interface ReceiverFormProps {
+  values: {
+    receiverName: string;
+    receiverPhone: string;
+    quantity: number;
+  };
+  errors: {
+    receiverName?: string;
+    receiverPhone?: string;
+    quantity?: string;
+  };
```
이 타입들은 기존의 타입을 재사용하거나 분리할 수 있어보여요 😊

shushuburger replied:
Pick과 Partial을 이용해서 미리 선언된 FormValues, FormErrors에서 필요한 부분만 가져와 쓸 수 있게 해보았습니다!

Leejin-Yang commented on src/hooks/useOrderForm.ts:
```diff
@@ -0,0 +1,132 @@
+    setFormValues(prev => ({ ...prev, [field]: value }));
+  };
+
+  const validateField = (field: FormField): boolean => {
```
이 함수는 좀 더 분리를 할 수 있어보여요!
field를 좀 더 활용해서 에러메시지를 set하는 함수와 에러메시지를 리턴하는 함수를 분리해서 사용해보면 어떨까요? 잠깐 든 생각으로는 에러메시지를 리턴하는 메서드들이 있는 객체를 만들어보는것도 좋아보이네요 😊

shushuburger replied:
우선은 함수를 분리하는 것에 초점을 맞추어 진행해보았습니다!! 조금 더 공부한 다음에 객체도 시도하도록 노력해보겠습니다!

Leejin-Yang reviewed (changes requested):
리뷰 잘 반영해주셨네요 👍

지금 마이페이지나 주문페이지의 경우 url을 직접 입력해 접속할 때 로그인 하지 않아도 접근이 가능한데, 이 부분만 수정 된다면 다음 단계로 넘어가도록 할게요!
나머지 코멘트들은 참고해주시고 반영할 부분 있다면 다음 단계에서 반영해주셔도 좋아보입니다 😊

확인해주시고 다시 요청 보내주세요 🙌

shushuburger committed:
feat(auth): isLoggedIn 상태를 AuthContext에 추가

- user !== null 여부 기반으로 로그인 상태를 제공하도록 했습니다.
- PrivateRoute 등에서 로그인 여부를 판별하여 /my와 /order/:id에 접근 여부를 가리도록 했습니다.

shushuburger committed:
refactor(login): 로그인 입력 필드 및 유효성 검사 로직 리팩토링

- LoginFormSection.tsx에서 타입 단언(as)을 제거하고 타입 가드 함수 isLoginField로 대체하였습니다.
- ===을 이용해 비교하도록 개선하였습니다.
- useLoginForm.ts에서 validateEmail, validatePassword 함수를 하나의 validateField(field) 함수로 통합해주었습니다.

shushuburger committed:
refactor(order): OrderPage 리팩토링

- handleInputChange에서 name as any 제거하고 isFormField 타입 가드 함수로 대체했습니다.
- FormField 타입을 useOrderForm.ts에서 export하여 외부에서도 안전하게 활용할 수 있도록 했습니다.
- navigate('/')를 navigate(ROUTES.HOME)으로 수정하여 하드코딩을 줄였습니다.

shushuburger committed:
refactor(order): validateField 로직 분리 및 getFieldError 함수 도입

- 각 필드 유효성 검사를 getFieldError 함수로 분리해주었습니다.
- validateField는 에러 메시지를 받아 setFormErrors만 처리하도록 분리했습니다.
- FormField 타입을 사용해주었습니다.

shushuburger committed:
refactor(order): ReceiverForm의 타입을 FormValues 기반으로 분리

- Pick과 Partial을 활용하여 중복된 타입 정의를 제거했습니다.

shushuburger committed:
fix(auth): 로그인 후 주소창에 입력하면 정상 이동하도록 수정

- AuthContext에 isLoading 상태 추가하여주었습니다.
- 로그인 후에 주소창에 /my를 입력하면 성공적으로 접속할 수 있도록 해주었습니다.

shushuburger commented:
로그인 시 /my, /order/2 성공적 접속
미로그인 시 /my, /order/2 입력하면 /login 으로 이동

위와 같이 동작하는 것 확인했습니다! npm run build 성공적으로 되는 것 확인하였습니다!

Leejin-Yang commented on src/components/LoginFormSection.tsx:
```diff
@@ -5,40 +5,47 @@ import { LOCATION_STATE_KEYS } from '@/constants/navigationState';
+  const { userInfo, handleChange, errors, validateField, isValidForm } =
+    useLoginForm();
+
+  const isLoginField = (name: string): name is 'email' | 'password' => {
+    return name === 'email' || name === 'password';
```
타입 가드 함수를 잘 작성해주셨네요 👍
이 함수는 컴포넌트 외부로 뺄 수 있겠네요 😊

Leejin-Yang commented on src/hooks/useOrderForm.ts:
```diff
@@ -0,0 +1,95 @@
+    setFormValues(prev => ({ ...prev, [field]: value }));
+  };
+
+  const validateField = (field: FormField): boolean => {
+    const value = formValues[field];
+    const error = getFieldError(field, value);
+
+    setFormErrors(prev => ({ ...prev, [field]: error }));
+    return !error;
+  };
```
유효성 검사가 한결 가벼워졌네요 👍👍
앞으로도 컴포넌트 작성이나 이런 로직들을 작성할 때, 선언형에 대해서 학습해보시면 좋아보입니다.

컴포넌트에서 커스텀 훅을 분리하는 것도 선언형의 일종이구요 😊

Leejin-Yang reviewed (approved):
이번 단계 잘 진행해주셔서 감사합니다 👍

앞으로도 지금처럼 컴포넌트나 함수가 복잡하지 않은지, 재사용과 분리에 대해 계속 고민해보면 좋아보여요 😊

이번 단계 고생 많으셨습니다. 다음 단계에서 보아요 🙌
